Source/SwiftLintFramework/Rules/Style/ImplicitSelfRule.swift

```diff
@@ -0,0 +1,231 @@
+        kind: .style
+    )
+
+    public func validate(file: SwiftLintFile) -> [StyleViolation] {
+        let identifiers = file.syntaxMap.tokens.compactMap { token -> Identifier? in
+            guard let kind = token.kind, kind == .identifier else { return nil }
+
+            return Identifier(token: token, file: file)
+        }
+
+        let structure = Structure(dict: file.structureDictionary.value)
+
+        for identifier in identifiers {
+            structure.add(identifier: identifier)
+        }
+
+        return violations(in: structure, declInScope: [], file: file)
+    }
+
+    private func violations(in structure: Structure,
+                            declInScope: Set<String>,
+                            file: SwiftLintFile) -> [StyleViolation] {
+        var existingDeclInScope: Set<String> = declInScope
+        guard !structure.isClosure else { return [] }
+        if structure.isInit && configuration.initSelfUsage == .always { return [] }
+        let violations = structure.identifiers.compactMap { identifier -> StyleViolation? in
+            guard identifier.name != "init" else { return nil }
+
+            guard !identifier.isParamDecl && !identifier.isLocalDecl else {
+                existingDeclInScope.insert(identifier.name)
+                return nil
+            }
+
+            if configuration.initSelfUsage == .beforeInitCall &&
+                structure.isInit &&
+                structure.hasInitBefore(identifier: identifier) {
+                return nil
+            }
+
+            let selfOffset = identifier.offset - 5
+            guard file.stringView.substringWithByteRange(ByteRange(location: selfOffset, length: 5)) == "self." else {
+                return nil
+            }
+
+            guard !existingDeclInScope.contains(identifier.name) else { return nil }
+
+            guard let range = file.stringView.byteRangeToNSRange(
+                ByteRange(location: selfOffset, length: 0)
+            ) else { return nil }
+
+            return StyleViolation(ruleDescription: type(of: self).description,
+                                  severity: configuration.severity.severity,
+                                  location: Location(file: file, characterOffset: range.location))
+        }
+
+        let subViolations = structure.substructure.flatMap { substructure -> [StyleViolation] in
+            if configuration.initSelfUsage == .beforeInitCall &&
+                structure.isInit &&
+                structure.hasInitBefore(structure: substructure) {
+                return []
+            }
+
+            return self.violations(in: substructure, declInScope: existingDeclInScope, file: file)
+        }
+
+        return violations + subViolations
+    }
+}
+
+private struct Identifier {
+    let name: String
+    let kind: String
+    let offset: ByteCount
+
+    init?(token: SwiftLintSyntaxToken, file: SwiftLintFile) {
```

Comment:
Why is this an optional initializer? Should we maybe return `nil` if `name` and/or `kind` would otherwise be initialized with a fallback value of `""`?